src/libraries/Common/src/System/Net/NTAuthentication.Managed.cs

```diff
@@ -0,0 +1,919 @@
+                        }
+                    }
+                }
+            }
+
+            return writer.Encode();
+        }
+
+        private unsafe byte[] ProcessSpNegoChallenge(byte[] challenge)
+        {
+            NegState state = NegState.Unknown;
+            string? mech = null;
+            byte[]? blob = null;
+
+            AsnReader reader = new AsnReader(challenge, AsnEncodingRules.DER);
+            AsnReader challengeReader = reader.ReadSequence(new Asn1Tag(TagClass.ContextSpecific, (int)NegotiationToken.NegTokenResp));
+            reader.ThrowIfNotEmpty();
+
+            // NegTokenResp::= SEQUENCE {
+            //    negState[0] ENUMERATED {
+            //        accept - completed(0),
+            //        accept - incomplete(1),
+            //        reject(2),
+            //        request - mic(3)
+            //    } OPTIONAL,
+            // --REQUIRED in the first reply from the target
+            //    supportedMech[1] MechType OPTIONAL,
+            // --present only in the first reply from the target
+            // responseToken[2] OCTET STRING  OPTIONAL,
+            // mechListMIC[3] OCTET STRING  OPTIONAL,
+            // ...
+            // }
+
+            challengeReader = challengeReader.ReadSequence();
+            while (challengeReader.HasData)
+            {
+                Asn1Tag tag = challengeReader.PeekTag();
+                if (tag.TagClass == TagClass.ContextSpecific)
+                {
+                    NegTokenResp dataType = (NegTokenResp)tag.TagValue;
+                    AsnReader specificValue = new AsnReader(challengeReader.PeekContentBytes(), AsnEncodingRules.DER);
+
+                    switch (dataType)
+                    {
+                        case NegTokenResp.NegState:
+                            state = specificValue.ReadEnumeratedValue<NegState>();
+                            specificValue.ThrowIfNotEmpty();
+                            break;
+                        case NegTokenResp.SupportedMech:
+                            mech = specificValue.ReadObjectIdentifier();
+                            specificValue.ThrowIfNotEmpty();
+                            break;
+                        case NegTokenResp.ResponseToken:
+                            blob = specificValue.ReadOctetString();
+                            specificValue.ThrowIfNotEmpty();
+                            break;
+                        default:
+                            // Ignore everything else
+                            break;
+                    }
+                }
+
+                challengeReader.ReadEncodedValue();
+            }
+
+            // Mechanism should be set on first message. That means always
+            // as NTLM has only one challenege message.
+            if (!NtlmOid.Equals(mech))
+            {
+                throw new Win32Exception(NTE_FAIL, $"'{mech}' mechanism is not supported");
```

Comment:
Exception messages should be in resx (and thus use format placeholders instead of interpolation)

Reply:
Yep, I have this one in the TODO in the initial comment.